File: python/92_square_digit_chains.py
def square_digit_chains(max_n: int) -> int:

    count: int = 0
    known = {89}
    
    for n in range(1, max_n + 1):
        result = n
        ends_at_89 = True
        chain = set()
        
        while True:
            result = sum([int(c) ** 2 for c in str(result)])
            chain.add(result)
            if result in known:
                known.update(chain)
                break
            if result == 1:
                ends_at_89 = False
                break
        
        if ends_at_89:
            count += 1
    
    return count
    
if __name__ == "__main__":
    print(square_digit_chains(10_000_000))

File: c#/_92_SquareDigitChains.cs
using System;
using System.Collections.Generic;

public class _92_SquareDigitChains{
	public static int Euler91(){
		List<int> nums = new List<int>();
		for (int i = 2; i < 10_000_000; i++){
			int n = i;
			while (n != 1 && n != 89){
				int sum = 0;
				while (n > 0){
					int d = n % 10;
					sum += d * d;
					n /= 10;
				}
				n = sum;
			}
			if (n == 89) nums.Add(i);
		}
		return nums.Count;
	}
	
	public static void Main(){
		Console.WriteLine(Euler91());
	}
}
